pkg/diskmaker/discovery/discovery.go

```diff
@@ -0,0 +1,262 @@
+		e := diskmaker.NewEvent(diskmaker.ErrorCreatingDiscoveryResultObject, fmt.Sprintf("%s. Error: %+v", message, err), "")
+		discovery.eventSync.Report(e, discovery.localVolumeDiscovery)
+		return errors.Wrapf(err, message)
+	}
+
+	err = discovery.discoverDevices()
+	if err != nil {
+		errors.Wrapf(err, "failed to discover devices")
+	}
+
+	// Watch udev events for continuous discovery of devices
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, syscall.SIGTERM)
+
+	udevEvents := make(chan string)
+	go udevBlockMonitor(udevEvents, udevEventPeriod)
+	for {
+		select {
+		case <-sigc:
+			klog.Info("shutdown signal received, exiting...")
+			return nil
+		case <-time.After(probeInterval):
+			if err := discovery.discoverDevices(); err != nil {
+				klog.Errorf("failed to discover devices during probe interval. %v", err)
+			}
+		case _, ok := <-udevEvents:
+			if ok {
+				klog.Info("trigger probe from udev event")
+				if err := discovery.discoverDevices(); err != nil {
+					klog.Errorf("failed to discover devices triggered from udev event. %v", err)
+				}
+			} else {
+				klog.Warningf("disabling udev monitoring")
+				udevEvents = nil
+			}
+		}
+	}
+}
+
+// discoverDevices identifies the list of usable disks on the current node
+func (discovery *DeviceDiscovery) discoverDevices() error {
+	// List all the valid block devices on the node
+	validDevices, err := getValidBlockDevices()
+	if err != nil {
+		message := "failed to discover devices"
+		e := diskmaker.NewEvent(diskmaker.ErrorListingBlockDevices, fmt.Sprintf("%s. Error: %+v", message, err), "")
+		discovery.eventSync.Report(e, discovery.localVolumeDiscovery)
+		return errors.Wrapf(err, message)
+	}
+
+	klog.Infof("valid block devices: %+v", validDevices)
+
+	discoveredDisks := getDiscoverdDevices(validDevices)
+	klog.Infof("discovered devices: %+v", discoveredDisks)
+
+	// Update discovered devices in the  LocalVolumeDiscoveryResult resource
+	if !reflect.DeepEqual(discovery.disks, discoveredDisks) {
+		klog.Info("device list updated. Updating LocalVolumeDiscoveryResult status...")
+		discovery.disks = discoveredDisks
+		err = discovery.updateStatus()
+		if err != nil {
+			message := "failed to update LocalVolumeDiscoveryResult status"
+			e := diskmaker.NewEvent(diskmaker.ErrorUpdatingDiscoveryResultObject, fmt.Sprintf("%s. Error: %+v", message, err), "")
+			discovery.eventSync.Report(e, discovery.localVolumeDiscovery)
+			return errors.Wrapf(err, message)
+		}
+		message := "successfully updated discovered device details in the LocalVolumeDiscoveryResult resource"
+		e := diskmaker.NewSuccessEvent(diskmaker.UpdatedDiscoveredDeviceList, message, "")
+		discovery.eventSync.Report(e, discovery.localVolumeDiscovery)
+	}
+
+	return nil
+}
+
+// getValidBlockDevices fetchs all the block devices sutitable for discovery
+func getValidBlockDevices() ([]internal.BlockDevice, error) {
+	blockDevices, badRows, err := internal.ListBlockDevices()
```

Comment:
We are ignoring and not listing LVM devices. I think this might be a problem. Was there an agreement about how to display LVM devices? i.e - if there are logical volumes out of a volume-group then are we going to list individual logical volumes or just the devices that make up the volume-group?

Reply:
We didn't have any discussions regarding the how to display LVM devices earlier. Also the deviceType in discovery is using same [deviceType](https://github.com/openshift/local-storage-operator/blob/2a671867f76c7b0d2e3ed0f89d1d55350b9b82f8/pkg/apis/local/v1alpha1/localvolumeset_types.go#L23) enum in auto provision which is restricted to options like `disk, part and loop`.

Reply:
@gnufied Can we have this change as part of bug fix. Because this will cause breaking changes in the UI and UI code is already merged.

Reply:
The problem is `deviceType` guarantee only exists for auto-provisioning API. It has a explicit field for that. Auto-discovery has no such restrictions and is expected to report most devices.